[SPARK-13923] [SQL] Implement SessionCatalog

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

As part of the effort to merge `SQLContext` and `HiveContext`, this patch implements an internal catalog called `SessionCatalog` that handles temporary functions and tables and delegates metastore operations to `ExternalCatalog`. Currently, this is still dead code, but in the future it will be part of `SessionState` and will replace `o.a.s.sql.catalyst.analysis.Catalog`.

A recent patch #11573 parses Hive commands ourselves in Spark, but still passes the entire query text to Hive. In a future patch, we will use `SessionCatalog` to implement the parsed commands.
## How was this patch tested?

800+ lines of tests in `SessionCatalogSuite`.

Commit:
Merge in @yhuai's changes

Commit:
Clean up table method signatures + add comments

Commit:
Do the same for functions and partitions

Commit:
Take into account current database in table methods

Commit:
Refactor CatalogTable to use TableIdentifier

This is a standalone commit such that in the future we can split
it out into a separate patch if preferrable.

Commit:
Refactor CatalogFunction to use FunctionIdentifier

Commit:
Fix tests

Commit:
Document and clean up function methods

Commit:
Implement SessionCatalog using ExternalCatalog

Commit:
Fix style

Commit:
Clean up duplicate code in Table/FunctionIdentifier

Commit:
Refactor CatalogTestCases to make methods accessible

Commit:
Fix infinite loop (woops)

Commit:
Implement tests for databases and tables

Commit:
Implement tests for table partitions

Commit:
Implement tests for functions

Commit:
Add TODO

SparkQA commented:
**[Test build #53263 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53263/consoleFull)** for PR 11750 at commit [`3b2e48a`](https://github.com/apache/spark/commit/3b2e48a4efa81bb886331ce5818e2d44c7c2f7da).

andrewor14 commented:
@yhuai @rxin

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -0,0 +1,484 @@
+   * If no such database is specified, create it in the current database.
+   */
+  def createTable(
+      currentDb: String,
```
it is somewhat strange to pass in currentDb and then rely on some table definition's database. Have you thought about just figuring out that part in the caller outside the session catalog? i.e. the catalog itself doesn't need to handle currentDb.

rxin replied:
As discussed offline, let's move the tracking of current db into SessionCatalog.

andrewor14 replied:
as discussed offline, this is because we need to deal with temporary tables. An alternative that I will implement here is to just keep track of the current database in this class so we don't need to pass it in everywhere.

rxin replied:
make sure we document that in the session catalog code and explain why we are tracking current db in here.

SparkQA commented:
**[Test build #53263 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53263/consoleFull)** for PR 11750 at commit [`3b2e48a`](https://github.com/apache/spark/commit/3b2e48a4efa81bb886331ce5818e2d44c7c2f7da).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53263/
Test FAILed.

Commit:
Keep track of current database in SessionCatalog

This allows us to not pass it into every single method like
we used to before this commit.

Commit:
Fix tests

Commit:
Expand test scope + clean up test code

SparkQA commented:
**[Test build #53358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53358/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -167,7 +170,7 @@ abstract class ExternalCatalog {
  * @param className fully qualified class name, e.g. "org.apache.spark.util.MyFunc"
  */
-case class CatalogFunction(name: String, className: String)
+case class CatalogFunction(name: FunctionIdentifier, className: String)
```
`catalogFunc.name.funcName` is kind of weird (we do not need to change it right now).

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -211,8 +214,7 @@ case class CatalogTablePartition(
 case class CatalogTable(
-    specifiedDatabase: Option[String],
-    name: String,
+    name: TableIdentifier,
```
Maybe we can rename this `name` later.

andrewor14 replied:
yeah, we should rename it. I think even something like `identifier` is better. Let's do that later.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -0,0 +1,469 @@
+   * If no database is specified, this will first attempt to drop a temporary table with
+   * the same name, then, if that does not exist, drop the table from the current database.
+   */
+  def dropTable(name: TableIdentifier, ignoreIfNotExists: Boolean): Unit = {
```
Just a note at here. If we have a temp table called `T1` and we also have a metastore in the current db called `T1`, this dropTable command will drop temp table `T1`.

yhuai replied:
This is the semantic of postgres and Hive.

rxin commented:
If there are no major problems, let's merge this as soon as tests pass. We can address minor comments as followups. This will unblock a bunch of other stuff.

yhuai commented:
Yea, let's merge this once it passes tests.

SparkQA commented:
**[Test build #2646 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2646/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).

SparkQA commented:
**[Test build #2647 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2647/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).

SparkQA commented:
**[Test build #2648 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2648/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).

nongli commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -0,0 +1,469 @@
+    if (tempTables.containsKey(name) && !ignoreIfExists) {
+      throw new AnalysisException(s"Temporary table '$name' already exists.")
+    }
+    tempTables.put(name, tableDefinition)
```
this isn't really ignoreIfExists but updateIfExists?

andrewor14 replied:
`ignoreIfExists` is like "ignoring the exception if the table already exists", but I guess it doesn't convey whether the table itself is overridden. I could rename this.

yhuai replied:
Let's make this name more explicit. I think in other places, `ignoreIfExists` means that if a table exists, we do nothing.

nongli commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -0,0 +1,469 @@
+   * Alter one or many table partitions whose specs that match those specified in `parts`,
+   * assuming the partitions exist.
+   *
+   * If no database is specified, assume the table is in the current database.
```
I don't think replicating this comment is that useful. This is just the semantics of current database.

nongli commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -0,0 +1,469 @@
+class SessionCatalog(externalCatalog: ExternalCatalog) {
+  import ExternalCatalog._
+
+  private[this] val tempTables = new ConcurrentHashMap[String, LogicalPlan]
```
Why concurrent? Comment on expected thread safety of this class.

yhuai replied:
I copied this from the existing catalog. Actually, I do not think multiple threads will use the same SessionCatalog. We can change it later.

nongli replied:
Ok. This thing is not thread safe right now and making these concurrent is not sufficient.

andrewor14 replied:
good catch. I just copied this from Yin's code. We can just make this a normal hash map then.

SparkQA commented:
**[Test build #53358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53358/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53358/
Test PASSed.

SparkQA commented:
**[Test build #2647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2647/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #2648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2648/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #2646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2646/consoleFull)** for PR 11750 at commit [`ad43a5f`](https://github.com/apache/spark/commit/ad43a5ffdeeb881aaed8944971b63a27d1f4257f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

yhuai commented:
OK. Let me merge this. Let's address comments in the follow-up pr.
